src/emailagent.cpp

```diff
@@ -42,6 +42,50 @@ QMailAccountId accountForMessageId(const QMailMessageId &msgId)
     QMailMessageMetaData metaData(msgId);
     return metaData.parentAccountId();
 }
+
+QString attachmentFilename(const QMailMessage &message,
+                           const QString &attachmentLocation)
+{
+    QString filename = message.customField(attachmentLocation + "-filename");
+    if (filename.isEmpty()) {
+        // Legacy naming scheme
+        const QMailMessagePartContainer::Location location(attachmentLocation);
+        if (message.contains(location)) {
+            filename = QStandardPaths::writableLocation(QStandardPaths::DownloadLocation)
+                + "/mail_attachments"
+                + "/" + QString::number(message.parentAccountId().toULongLong())
+                + "/" + attachmentLocation
+                + "/" + message.partAt(location).displayName().remove('/');
+        }
+    }
+    return filename;
+}
+
+void setAttachmentFilename(QMailMessageMetaData *meta,
+                           const QString &attachmentLocation,
+                           const QString &filename)
+{
+    meta->setCustomField(attachmentLocation + "-filename", filename);
+    QMailMessageMetaData local = *meta;
+    QTimer::singleShot(0, [local] {
+                              QMailStore::instance()->updateMessage((QMailMessageMetaData*)&local);
+                          });
+}
+
+QString attachmentPath(const QMailMessageContentType &type)
+{
+    if (type.matches("image")) {
+        return QStandardPaths::writableLocation(QStandardPaths::PicturesLocation);
```

Comment:
I'm thinking we should consider all attachments as downloads, stored to the generic downloads location. User can then check what he/she has downloaded from a single place.

Reply:
I thought this change may be controversial. That's why I put it in a separated commit. I can change it to use the top level of ~/Downloads. Personally, I prefer that files get well tidy in a semi automatic way. So I don't have to move them later on by hand. But anyway no problem to change this if there is no consensus.

Reply:
Moving elsewhere might be good for some cases but confusing for other. E.g. user gets a pdf as attachment, saves it, then decides to send it forward to someone else, on file picker assumes it's found under downloads. For comparison browser doesn't put different types on different directories either.

Reply:
I still prefer when files are tidy, but I understand why you prefer to put everything in download. No problem to stay with the latter. Besides, my solution was depend on the right MIME type, which is fragile and out of our hands. I received an email with a PDF attachment given as a `application/x-download`. Obviously with my solution it didn't end up in `~/Documents` which was very misleading.